# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [99]:
# Your code here
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [100]:
# read dataframe using read_csv(url)
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')


# checking the dataframe, column names, data types

display(df.head())

display(df.tail()) # output shows many NaN values at the bottom of the dataset

# checking dimension of df, number of rows and columns
display(df.shape)

# checking general infomation  of the data, columns names, non-null count, dtyoe
df.info()


Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

Customer   ST GENDER Education Customer Lifetime Value  Income  \
4003      NaN  NaN    NaN       NaN                     NaN     NaN   
4004      NaN  NaN    NaN       NaN                     NaN     NaN   
4005      NaN  NaN    NaN       NaN                     NaN     NaN   
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
4003                   NaN                       NaN         NaN   
4004                   NaN                       NaN         NaN   
4005                   NaN                       NaN         NaN   
4006                   NaN                       NaN         NaN   
4007                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
4003           NaN                 NaN  
4004           NaN                 NaN  
4005           NaN                 NaN  
4006           NaN                 NaN  
4007           NaN                 NaN

(4008, 11)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [101]:
# convert colname into lowercase, and replace " " with  "_", rename 'st' to 'state

cols=[]

for c in df.columns:
    cols.append(c.lower().replace(" ", "_"))
df.columns = cols

df = df.rename(columns = {'st' : 'state'})

df

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [102]:
# Your code here

# Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".

df.loc[df['gender'] == 'Male', 'gender' ] = 'M'

df.loc[df['gender'] .isin (['female', 'Femal']), 'gender' ] = 'F'

display(df['gender'].value_counts())



F    502
M    452
Name: gender, dtype: int64

In [103]:
# functions to clean 'gender' 
def gender_clean (df:pd.DataFrame) ->pd.DataFrame:
    '''
    This function will get the 'gender' column (which contain string datatype) in the provided dataframe.
    Then convert it into lower case.
    Then get the first letter, and check whether it is 'm' then return 'M'
    , or 'f' then return 'F'
    , or else return 'U' as unknow
    
    Input:
    df: Pandas DataFrame
    
    Output:
    A Pandas DataFrame with a cleaned value of column 'gender'. 
    Gender values: 'F', 'M', 'U'
    '''
    df2 = df.copy()
    if 'gender' not in df2.columns:
        return df2
    else:
        df2['gender']= df2['gender'].apply(lambda x: x[0].upper() if x[0].upper() in ['M','F'] else 'U')
        return df2
#df[gender_column] = df[gender_column].apply(lambda x: 'F' if x.lower() == 'f' else ('M' if x.lower() == 'm' else 'U'))


In [104]:
# State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"

display(df['state'].value_counts())

''''
# one way to do it
df.loc[df['state'] == 'Cali', 'state' ] = 'California'
df.loc[df['state'] == 'AZ', 'state' ] = 'Arizona'
df.loc[df['state'] == 'WA', 'state' ] = 'Washington'
display(df['state'].value_counts())
'''

# another way to do it:
df['state'] = df['state'].replace({'Cali': 'California', 'AZ': 'Arizona', 'WA': 'Washington'})
display(df['state'].value_counts())


Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: state, dtype: int64

California    331
Oregon        320
Arizona       211
Washington    111
Nevada         98
Name: state, dtype: int64

In [105]:
# In education, "Bachelors" could be replaced by "Bachelor"
df.loc[df['education'] == 'Bachelors', 'education' ] = 'Bachelor'
display(df['education'].value_counts())

# OR
df['education'] = df['education'].replace({'Bachelors': 'Bachelor'})
display(df['education'].value_counts())

#In Customer Lifetime Value, delete the % character

df['customer_lifetime_value'] = df['customer_lifetime_value'].replace('%', '')
display(df)

# Question: for all above columns, when I used df.replace() directly, I could 
# replace specific string/characters in the column. But for 'customer_lifetime_value', using 
# df.['customer_lifetime_value'].replace('%', '') --> did not do anything, as I still can see '%'
# in the output. AND customer_lifetime_value dtype is object
# And for it to work, I need to use df.['customer_lifetime_value'].str.replace('%', '') as shown in the next cell,
# and it works???????


#In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"
df['vehicle_class'] = df['vehicle_class'].apply(lambda x: 'Luxury' if x in ['Sports Car', 'Luxury SUV', 'Luxury Car'] else x) 
display(df['vehicle_class'].value_counts())

Bachelor                331
College                 313
High School or Below    296
Master                   94
Doctor                   37
Name: education, dtype: int64

Bachelor                331
College                 313
High School or Below    296
Master                   94
Doctor                   37
Name: education, dtype: int64

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

Four-Door Car    576
Two-Door Car     205
SUV              199
Luxury            91
Name: vehicle_class, dtype: int64

## NOTE: 
I think i might have the answer for the question regarding dtype('O'), and df.replace vs df.str.replace
But I still would like to hear from you the answer, to see whether my research for the answer correct or not. Thank you!

In [106]:
# WHY: df.str.replace() works in replacing '%' but not df.replace()
# as dtype of 'customer_lifetie_value' is object (string), so is it a must still to have str. before replace()
# to be able to apply pd.string replace method???

df['customer_lifetime_value'] = df['customer_lifetime_value'].str.replace('%', '')
display(df)
df['gender'].dtype
# also, I am not entirely understand why, when I use df.dtypes
# 'customer_lifetime_value' show dtype: object
# but if I use df['customer_lifetime_value'].dtype
# the output is dtype('0')??? which is a strange dtype???


customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                  697953.59      0.0                  94.0   
2                 1288743.17  48767.0                 108.0   
3                  764586.18      0.0                 106.0   
4                  536307.65  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

dtype('O')

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [107]:
# Your code here

# converting customer_lifetime_value to numeric

df['customer_lifetime_value'] = pd.to_numeric(df['customer_lifetime_value'], errors = 'coerce')

display(df.dtypes)
df

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value      float64
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
...                       ...      ...                   ...   
4003                      NaN      NaN                   NaN   
4004                      NaN      NaN                   NaN   
4005                      NaN      NaN                   NaN   
4006                      NaN      NaN                   NaN   
4007                      NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [108]:
# check the unique values 
display(df['number_of_open_complaints'].unique())

# correcting number_open_complaints format, 1/5/00 should be 5

display(df['number_of_open_complaints'].value_counts()) # checking counts of unique values before correcting

df['number_of_open_complaints'] = pd.to_numeric(df['number_of_open_complaints'].str.split('/').str[1])

display(df['number_of_open_complaints'].value_counts()) # check counts and unique values after correcting

df


array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan],
      dtype=object)

1/0/00    830
1/1/00    138
1/2/00     50
1/3/00     34
1/4/00     13
1/5/00      6
Name: number_of_open_complaints, dtype: int64

0.0    830
1.0    138
2.0     50
3.0     34
4.0     13
5.0      6
Name: number_of_open_complaints, dtype: int64

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
...                       ...      ...                   ...   
4003                      NaN      NaN                   NaN   
4004                      NaN      NaN                   NaN   
4005                      NaN      NaN                   NaN   
4006                      NaN      NaN                   NaN   
4007                      NaN      NaN                   NaN   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
4003                        NaN             NaN            NaN   
4004                        NaN             NaN            NaN   
4005                        NaN             NaN            NaN   
4006                        NaN             NaN            NaN   
4007                        NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers using `applymap()`.

In [109]:
# Your code here

# checking na value in the dataset

display(df.isnull())

display(df.isnull().sum())
# output shows more than half of dataset entries have missing values, and in all columns

# checking missing values percentage in df (round 1)
missing_values_df = pd.DataFrame(round(df.isnull().sum()/len(df),4)*100)
display(missing_values_df)
missing_values_df = missing_values_df.reset_index()
display(missing_values_df)
missing_values_df.columns = ['column_name', 'percentage_of_missing_values']
display(missing_values_df)

customer  state  gender  education  customer_lifetime_value  income  \
0        False  False    True      False                     True   False   
1        False  False   False      False                    False   False   
2        False  False   False      False                    False   False   
3        False  False   False      False                    False   False   
4        False  False   False      False                    False   False   
...        ...    ...     ...        ...                      ...     ...   
4003      True   True    True       True                     True    True   
4004      True   True    True       True                     True    True   
4005      True   True    True       True                     True    True   
4006      True   True    True       True                     True    True   
4007      True   True    True       True                     True    True   

      monthly_premium_auto  number_of_open_complaints  policy_type  \
0                    False                      False        False   
1                    False                      False        False   
2                    False                      False        False   
3                    False                      False        False   
4                    False                      False        False   
...                    ...                        ...          ...   
4003                  True                       True         True   
4004                  True                       True         True   
4005                  True                       True         True   
4006                  True                       True         True   
4007                  True                       True         True   

      vehicle_class  total_claim_amount  
0             False               False  
1             False               False  
2             False               False  
3             False               False  
4             False               False  
...             ...                 ...  
4003           True                True  
4004           True                True  
4005           True                True  
4006           True                True  
4007           True                True  

[4008 rows x 11 columns]

customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

0
customer                   73.28
state                      73.28
gender                     76.20
education                  73.28
customer_lifetime_value    73.35
income                     73.28
monthly_premium_auto       73.28
number_of_open_complaints  73.28
policy_type                73.28
vehicle_class              73.28
total_claim_amount         73.28

index      0
0                    customer  73.28
1                       state  73.28
2                      gender  76.20
3                   education  73.28
4     customer_lifetime_value  73.35
5                      income  73.28
6        monthly_premium_auto  73.28
7   number_of_open_complaints  73.28
8                 policy_type  73.28
9               vehicle_class  73.28
10         total_claim_amount  73.28

column_name  percentage_of_missing_values
0                    customer                         73.28
1                       state                         73.28
2                      gender                         76.20
3                   education                         73.28
4     customer_lifetime_value                         73.35
5                      income                         73.28
6        monthly_premium_auto                         73.28
7   number_of_open_complaints                         73.28
8                 policy_type                         73.28
9               vehicle_class                         73.28
10         total_claim_amount                         73.28

In [110]:
# after calculate the percentage of missing values, it shows more than 70% of data has missing values
# However, considering the number of missing values in all columns
# it suggests that minimum 2937 entries in this data set are empty entries, with missing values for all columns

# I check the bottom 2937 entries of the dataset
# after seeing that it seems these entries are empty, I create a new df called
# bottom_2937_df to contain these entries for further check

bottom_2937_df= df.tail(2937)
#bottom_2937_df= df.iloc[1071:4007,]
display(bottom_2937_df)
display(bottom_2937_df.isnull().sum()) # confirmed these are empty entries

# I will just drop these bottom 2937 entries, as they are all empty
# and base on any column that has in total only 2937 NA values
df = df[df['policy_type'].isnull() == False]
display(df)

display(df.isnull().sum())

customer state gender education  customer_lifetime_value  income  \
1071      NaN   NaN    NaN       NaN                      NaN     NaN   
1072      NaN   NaN    NaN       NaN                      NaN     NaN   
1073      NaN   NaN    NaN       NaN                      NaN     NaN   
1074      NaN   NaN    NaN       NaN                      NaN     NaN   
1075      NaN   NaN    NaN       NaN                      NaN     NaN   
...       ...   ...    ...       ...                      ...     ...   
4003      NaN   NaN    NaN       NaN                      NaN     NaN   
4004      NaN   NaN    NaN       NaN                      NaN     NaN   
4005      NaN   NaN    NaN       NaN                      NaN     NaN   
4006      NaN   NaN    NaN       NaN                      NaN     NaN   
4007      NaN   NaN    NaN       NaN                      NaN     NaN   

      monthly_premium_auto  number_of_open_complaints policy_type  \
1071                   NaN                        NaN         NaN   
1072                   NaN                        NaN         NaN   
1073                   NaN                        NaN         NaN   
1074                   NaN                        NaN         NaN   
1075                   NaN                        NaN         NaN   
...                    ...                        ...         ...   
4003                   NaN                        NaN         NaN   
4004                   NaN                        NaN         NaN   
4005                   NaN                        NaN         NaN   
4006                   NaN                        NaN         NaN   
4007                   NaN                        NaN         NaN   

     vehicle_class  total_claim_amount  
1071           NaN                 NaN  
1072           NaN                 NaN  
1073           NaN                 NaN  
1074           NaN                 NaN  
1075           NaN                 NaN  
...            ...                 ...  
4003           NaN                 NaN  
4004           NaN                 NaN  
4005           NaN                 NaN  
4006           NaN                 NaN  
4007           NaN                 NaN  

[2937 rows x 11 columns]

customer                     2937
state                        2937
gender                       2937
education                    2937
customer_lifetime_value      2937
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327  California      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
...                       ...      ...                   ...   
1066                305955.03  38644.0                  78.0   
1067               2031499.76  63209.0                 102.0   
1068                323912.47  16061.0                  88.0   
1069                462680.11  79487.0                 114.0   
1070                899704.02  54230.0                 112.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
1066                        1.0   Personal Auto  Four-Door Car   
1067                        2.0   Personal Auto            SUV   
1068                        0.0   Personal Auto  Four-Door Car   
1069                        0.0    Special Auto            SUV   
1070                        0.0   Personal Auto   Two-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[1071 rows x 11 columns]

customer                       0
state                          0
gender                       117
education                      0
customer_lifetime_value        3
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

In [111]:
# Now I am checking for left over missing value in the new df (round 2)

missing_values_df_2 = pd.DataFrame(round(df.isnull().sum()/len(df),4)*100)
display(missing_values_df_2)
missing_values_df_2 = missing_values_df_2.reset_index()
display(missing_values_df_2)
missing_values_df_2.columns = ['column_name', 'percentage_of_missing_values']
display(missing_values_df_2)

# Now based on round 2 analysis of missing values,
# I can see that column 'gender' has 10.92% entries with missing values 
# and column 'customer_lifetime_value' has 0.28% entries with missing values
# therefore, I can go ahead with dropping entries has na value in column 'customer_lifetime_value'

df = df[df['customer_lifetime_value'].isnull() == False]
display(df)
display(df.isnull().sum())

df['gender'].value_counts() # checking value counts of column 'gender'

# since we could not know about the gender of the customer with na value, 
# considering the value_counts of 'F', 'M', if we fill na with most frequent value, in this case 'F'
# it will introduce a big bias in data regarding gender
# as it will add 116 entries with F, which equal 20% of current F entries or 25% of current 'M' entries

#therefore, I will make a new df that contains these entries with na value of gender:
gender_null_116_df = df[df['gender'].isnull() == True]
display(gender_null_116_df)
# and update df without 116 entries with missing value in gender column
# by doing so, you can go back and combine again these 2 datasets (gender_null_116_df and df_clean) when needed
df_clean = df[df['gender'].isnull() == False]
display(df_clean)

display(df_clean.isnull().sum())


0
customer                    0.00
state                       0.00
gender                     10.92
education                   0.00
customer_lifetime_value     0.28
income                      0.00
monthly_premium_auto        0.00
number_of_open_complaints   0.00
policy_type                 0.00
vehicle_class               0.00
total_claim_amount          0.00

index      0
0                    customer   0.00
1                       state   0.00
2                      gender  10.92
3                   education   0.00
4     customer_lifetime_value   0.28
5                      income   0.00
6        monthly_premium_auto   0.00
7   number_of_open_complaints   0.00
8                 policy_type   0.00
9               vehicle_class   0.00
10         total_claim_amount   0.00

column_name  percentage_of_missing_values
0                    customer                          0.00
1                       state                          0.00
2                      gender                         10.92
3                   education                          0.00
4     customer_lifetime_value                          0.28
5                      income                          0.00
6        monthly_premium_auto                          0.00
7   number_of_open_complaints                          0.00
8                 policy_type                          0.00
9               vehicle_class                          0.00
10         total_claim_amount                          0.00

customer       state gender             education  \
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327  California      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

      customer_lifetime_value   income  monthly_premium_auto  \
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
5                   825629.78  62902.0                  69.0   
...                       ...      ...                   ...   
1066                305955.03  38644.0                  78.0   
1067               2031499.76  63209.0                 102.0   
1068                323912.47  16061.0                  88.0   
1069                462680.11  79487.0                 114.0   
1070                899704.02  54230.0                 112.0   

      number_of_open_complaints     policy_type  vehicle_class  \
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
5                           0.0   Personal Auto   Two-Door Car   
...                         ...             ...            ...   
1066                        1.0   Personal Auto  Four-Door Car   
1067                        2.0   Personal Auto            SUV   
1068                        0.0   Personal Auto  Four-Door Car   
1069                        0.0    Special Auto            SUV   
1070                        0.0   Personal Auto   Two-Door Car   

      total_claim_amount  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
5             159.383042  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[1068 rows x 11 columns]

customer                       0
state                          0
gender                       116
education                      0
customer_lifetime_value        0
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

customer       state gender             education  \
11   VQ65197  California    NaN               College   
12   DP39365  California    NaN                Master   
13   SJ95423     Arizona    NaN  High School or Below   
14   IL66569  California    NaN               College   
15   BW63560      Oregon    NaN              Bachelor   
..       ...         ...    ...                   ...   
134  BU44523     Arizona    NaN  High School or Below   
135  MS41162     Arizona    NaN  High School or Below   
136  PU25891      Oregon    NaN  High School or Below   
138  SX26335      Oregon    NaN               College   
449  LN31673  Washington    NaN               College   

     customer_lifetime_value   income  monthly_premium_auto  \
11                 819719.71      0.0                 110.0   
12                 879879.70  77026.0                 110.0   
13                 881901.89  99845.0                 110.0   
14                 538443.17  83689.0                  70.0   
15                 746313.94  24599.0                  64.0   
..                       ...      ...                   ...   
134                802637.96      0.0                  74.0   
135                402296.35      0.0                 117.0   
136                961831.09  80536.0                 119.0   
138                955292.69  97732.0                  79.0   
449               1131808.98  38923.0                  99.0   

     number_of_open_complaints     policy_type  vehicle_class  \
11                         0.0   Personal Auto            SUV   
12                         2.0  Corporate Auto  Four-Door Car   
13                         1.0  Corporate Auto            SUV   
14                         2.0  Corporate Auto  Four-Door Car   
15                         1.0  Corporate Auto  Four-Door Car   
..                         ...             ...            ...   
134                        0.0  Corporate Auto  Four-Door Car   
135                        0.0   Personal Auto            SUV   
136                        0.0    Special Auto         Luxury   
138                        0.0   Personal Auto  Four-Door Car   
449                        2.0   Personal Auto  Four-Door Car   

     total_claim_amount  
11           528.000000  
12           472.029737  
13           528.000000  
14           307.139132  
15            42.920271  
..                  ...  
134          532.800000  
135          975.107098  
136           53.798708  
138          289.912200  
449          115.728852  

[116 rows x 11 columns]

customer       state gender             education  \
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327  California      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

      customer_lifetime_value   income  monthly_premium_auto  \
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
5                   825629.78  62902.0                  69.0   
...                       ...      ...                   ...   
1066                305955.03  38644.0                  78.0   
1067               2031499.76  63209.0                 102.0   
1068                323912.47  16061.0                  88.0   
1069                462680.11  79487.0                 114.0   
1070                899704.02  54230.0                 112.0   

      number_of_open_complaints     policy_type  vehicle_class  \
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
5                           0.0   Personal Auto   Two-Door Car   
...                         ...             ...            ...   
1066                        1.0   Personal Auto  Four-Door Car   
1067                        2.0   Personal Auto            SUV   
1068                        0.0   Personal Auto  Four-Door Car   
1069                        0.0    Special Auto            SUV   
1070                        0.0   Personal Auto   Two-Door Car   

      total_claim_amount  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
5             159.383042  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[952 rows x 11 columns]

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [112]:
# After formatting data types
#as a last step, convert all the numeric variables to integers using applymap()

display(df_clean.dtypes)

display(df_clean.select_dtypes('float64'))
num_var = ['customer_lifetime_value', 'income', 'monthly_premium_auto', 'number_of_open_complaints', 'total_claim_amount']
df_clean = df_clean.applymap(lambda x: int(x) if isinstance(x, (float,int)) else x)
display(df_clean.dtypes)

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value      float64
income                       float64
monthly_premium_auto         float64
number_of_open_complaints    float64
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

customer_lifetime_value   income  monthly_premium_auto  \
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
5                   825629.78  62902.0                  69.0   
...                       ...      ...                   ...   
1066                305955.03  38644.0                  78.0   
1067               2031499.76  63209.0                 102.0   
1068                323912.47  16061.0                  88.0   
1069                462680.11  79487.0                 114.0   
1070                899704.02  54230.0                 112.0   

      number_of_open_complaints  total_claim_amount  
1                           0.0         1131.464935  
2                           0.0          566.472247  
3                           0.0          529.881344  
4                           0.0           17.269323  
5                           0.0          159.383042  
...                         ...                 ...  
1066                        1.0          361.455219  
1067                        2.0          207.320041  
1068                        0.0          633.600000  
1069                        0.0          547.200000  
1070                        0.0          537.600000  

[952 rows x 5 columns]

customer                     object
state                        object
gender                       object
education                    object
customer_lifetime_value       int64
income                        int64
monthly_premium_auto          int64
number_of_open_complaints     int64
policy_type                  object
vehicle_class                object
total_claim_amount            int64
dtype: object

## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [117]:
# Your code here

df_clean.duplicated().sum()

# my output for df_clean.duplicated().sum() is 0
# I don't have duplicates in the df_clean dataset ?????

# reset_index of df_clean
df_clean = df_clean.reset_index()

df_clean = df_clean.drop(columns = 'index')
display(df_clean)
display(df_clean.columns)

df_clean.to_csv('customer_data_insurance_comp.csv', index = False)

!ls

customer       state gender             education  \
0    QZ44356     Arizona      F              Bachelor   
1    AI49188      Nevada      F              Bachelor   
2    WW63253  California      M              Bachelor   
3    GA49547  Washington      M  High School or Below   
4    OC83172      Oregon      F              Bachelor   
..       ...         ...    ...                   ...   
947  TM65736      Oregon      M                Master   
948  VJ51327  California      F  High School or Below   
949  GS98873     Arizona      F              Bachelor   
950  CW49887  California      F                Master   
951  MY31220  California      F               College   

     customer_lifetime_value  income  monthly_premium_auto  \
0                     697953       0                    94   
1                    1288743   48767                   108   
2                     764586       0                   106   
3                     536307   36357                    68   
4                     825629   62902                    69   
..                       ...     ...                   ...   
947                   305955   38644                    78   
948                  2031499   63209                   102   
949                   323912   16061                    88   
950                   462680   79487                   114   
951                   899704   54230                   112   

     number_of_open_complaints     policy_type  vehicle_class  \
0                            0   Personal Auto  Four-Door Car   
1                            0   Personal Auto   Two-Door Car   
2                            0  Corporate Auto            SUV   
3                            0   Personal Auto  Four-Door Car   
4                            0   Personal Auto   Two-Door Car   
..                         ...             ...            ...   
947                          1   Personal Auto  Four-Door Car   
948                          2   Personal Auto            SUV   
949                          0   Personal Auto  Four-Door Car   
950                          0    Special Auto            SUV   
951                          0   Personal Auto   Two-Door Car   

     total_claim_amount  
0                  1131  
1                   566  
2                   529  
3                    17  
4                   159  
..                  ...  
947                 361  
948                 207  
949                 633  
950                 547  
951                 537  

[952 rows x 11 columns]

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

README.md
cleanfunctions.py
custome_data_insurance_comp.csv
customer_data_insurance_comp.csv
lab-dw-data-cleaning-and-formatting.ipynb


# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [58]:
%%writefile cleanfunctions.py

import pandas as pd

def lower_case_no_space_colname (df:pd.DataFrame)->pd.DataFrame:
    '''
    This function takes a Pandas DataFrame as input, and created an interal copy.
    Then on the internal copy, it will convert all column names to lower case and replace white space with '_'.
    
    Input:
    df:Pandas DataFrame
        
    Output:
    Modified Pandas DataFrame
    
    '''
    df2 = df.copy()
    cols = []
    for c in df2.columns:
        cols.append(c.lower().replace(' ', '_'))
    df2.columns = cols
    return df2

def delete_unamed_column(df:pd.DataFrame, unnamed_col = ['Unnamed: 0', 'Unnamed:_0']) -> pd.DataFrame:
    '''
    This function takes a Pandas DataFrame as input, and creates an internal copy.
    Then on the interal copy, it checks whether or not contains a column named "Unnamed: 0" or "Unnamed:_0".
    If it exists, deletes the column and returns the modified DataFrame, otherwise returns
    the original DataFrame.

    Inputs:
    df: Pandas DataFrame

    Output:
    Modified Pandas DataFrame
    '''

    df2 = df.copy()

    for unnamed_col in df2.columns:
        df2 = df2.drop(unnamed_col, axis=1)

    return df2

def clean_gender_column(df:pd.DataFrame) -> pd.DataFrame:
    '''
    This function will take a Pandas DataFrame as an input and it will replace the values in
    the "gender" column ins such a way that any gender which is not Male or Female with be 
    replaced by "U" otherwise the genders will be either "F" or "M"

    Inputs:
    df: Pandas DataFrame

    Outputs:
    A pandas DataFrame with the values in the "gender" column cleaned.
    '''

    df2 = df.copy()

    if "gender" not in df2.columns:
        return df2
    else:
        df2['gender'] = df2['gender'].apply(lambda x: x[0].upper() if x[0].upper() in ['M', 'F'] else "U")
        #df2['gender'] = list(map(lambda x: x[0].upper() if x[0].upper() in ['M', 'F'] else "U", df2['gender']))
        return df2

        
def checking_null_value (df:pd.DataFrame, threshold = 3)-> pd.DataFrame:
    '''
    This function takes a Pandas DataFrame, calculate the percentage of null value
    present in each column. 
    Return one dataframe showing a complete columns with their percentage of null value,
    and another dataframe showing only comlumns which have a percentage of null value
    above the input threshold.
    
    
    Input:
    df:Pandas DataFrame
    an optional threshold value: numeric (float, int)
    
    Output:
    2 Pandas DataFrame: one shows percentage of null value in all columns.
    And the other shows columns with null value percentage above the provided threshold
    '''
    
    missing_values_df = pd.DataFrame(round(df.isna().sum()/len(df),4)*100)

    missing_values_df = missing_values_df.reset_index()

    missing_values_df.columns = ['column_name', 'percentage_of_missing_values']
    column_with_null_higher_than_threshold_df = missing_values_df[missing_values_df['percentage_of_missing_values']>threshold]
    return  missing_values_df
    return  column_with_null_higher_than_threshold_df      
    

def clean_dataframe(df: pd.DataFrame, options: dict={"column_name": "state"}) -> pd.DataFrame:
    '''
    This function will take a Pandas DataFrame and it will apply the previous functions in the library
    to clean some columns of the dataframe

    Inputs: 
    df: Pandas DataFrame

    Outputs:
    Another DataFrame
    '''

    df2 = df.copy()

    df2 = lower_case_no_space_colname(df2)
    df2 = delete_unamed_column(df2)
    df2 = clean_gender_column(df2)

    return df2

Overwriting cleanfunctions.py


In [59]:
!ls

README.md
cleanfunctions.py
custome_data_insurance_comp.csv
customer_data_insurance_comp.csv
lab-dw-data-cleaning-and-formatting.ipynb


In [8]:
#test functions code
import pandas as pd

df_test = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
df_test

def lower_case_no_space_colname (df:pd.DataFrame)->pd.DataFrame:
    '''
    This function will convert all column names to lower case and replace white space with '_'
    
    Input:
    df:Pandas DataFrame
        
    Output:
    df:Pandas DataFrame
    
    '''
    df2 = df.copy()
    cols = []
    for c in df2.columns:
        cols.append(c.lower().replace(' ', '_'))
    df2.columns = cols
    return df2

lower_case_no_space_colname(df_test)
display(df_test)
display(df2)

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

customer          st gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [135]:
# Your code here
df_clean.info()
df_clean.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 952 entries, 0 to 951
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   customer                   952 non-null    object
 1   state                      952 non-null    object
 2   gender                     952 non-null    object
 3   education                  952 non-null    object
 4   customer_lifetime_value    952 non-null    int64 
 5   income                     952 non-null    int64 
 6   monthly_premium_auto       952 non-null    int64 
 7   number_of_open_complaints  952 non-null    int64 
 8   policy_type                952 non-null    object
 9   vehicle_class              952 non-null    object
 10  total_claim_amount         952 non-null    int64 
dtypes: int64(5), object(6)
memory usage: 81.9+ KB


customer_lifetime_value        income  monthly_premium_auto  \
count             9.520000e+02    952.000000            952.000000   
mean              7.942008e+05  39557.350840            205.231092   
std               6.258791e+05  30744.561851           1697.752504   
min               2.004350e+05      0.000000             61.000000   
25%               4.117828e+05  13087.750000             68.000000   
50%               5.885740e+05  36632.000000             83.000000   
75%               9.037502e+05  65085.500000            110.000000   
max               4.479546e+06  99960.000000          35354.000000   

       number_of_open_complaints  total_claim_amount  
count                 952.000000          952.000000  
mean                    0.378151          404.544118  
std                     0.872446          299.252620  
min                     0.000000            0.000000  
25%                     0.000000          196.000000  
50%                     0.000000          350.000000  
75%                     0.000000          533.250000  
max                     5.000000         2893.000000

In [130]:
potential_areas_threshold = df_clean[['customer_lifetime_value','total_claim_amount']].describe()
potential_areas_threshold

customer_lifetime_value  total_claim_amount
count             9.520000e+02          952.000000
mean              7.942008e+05          404.544118
std               6.258791e+05          299.252620
min               2.004350e+05            0.000000
25%               4.117828e+05          196.000000
50%               5.885740e+05          350.000000
75%               9.037502e+05          533.250000
max               4.479546e+06         2893.000000

In [133]:
#Create a pandas DataFrame object that contains information about 
#customers with a policy claim amount greater than the 75th percentile
#and a customer lifetime value in the bottom 25th percentile.

potential_areas_df = df_clean[(df_clean['total_claim_amount']>potential_areas_threshold.iloc[6,1]) | (df_clean['customer_lifetime_value']<potential_areas_threshold.iloc[4,0])]
display(potential_areas_df)


customer       state gender             education  \
0    QZ44356     Arizona      F              Bachelor   
1    AI49188      Nevada      F              Bachelor   
20   HS28694  Washington      F  High School or Below   
23   DS81757      Oregon      M               College   
25   LP84436  California      F  High School or Below   
..       ...         ...    ...                   ...   
946  XD85577  California      M              Bachelor   
947  TM65736      Oregon      M                Master   
949  GS98873     Arizona      F              Bachelor   
950  CW49887  California      F                Master   
951  MY31220  California      F               College   

     customer_lifetime_value  income  monthly_premium_auto  \
0                     697953       0                    94   
1                    1288743   48767                   108   
20                    282986   25317                    71   
23                    247012       0                    74   
25                    904711       0                   127   
..                       ...     ...                   ...   
946                   284226   69417                    73   
947                   305955   38644                    78   
949                   323912   16061                    88   
950                   462680   79487                   114   
951                   899704   54230                   112   

     number_of_open_complaints    policy_type  vehicle_class  \
0                            0  Personal Auto  Four-Door Car   
1                            0  Personal Auto   Two-Door Car   
20                           0  Personal Auto   Two-Door Car   
23                           0  Personal Auto   Two-Door Car   
25                           0  Personal Auto         Luxury   
..                         ...            ...            ...   
946                          1  Personal Auto  Four-Door Car   
947                          1  Personal Auto  Four-Door Car   
949                          0  Personal Auto  Four-Door Car   
950                          0   Special Auto            SUV   
951                          0  Personal Auto   Two-Door Car   

     total_claim_amount  
0                  1131  
1                   566  
20                   50  
23                  721  
25                 1087  
..                  ...  
946                  30  
947                 361  
949                 633  
950                 547  
951                 537  

[441 rows x 11 columns]

In [134]:
potential_areas_df[['total_claim_amount', 'customer_lifetime_value']].describe()

total_claim_amount  customer_lifetime_value
count           441.00000             4.410000e+02
mean            544.46712             6.216396e+05
std             361.85645             5.488525e+05
min               1.00000             2.004350e+05
25%             312.00000             2.725350e+05
50%             546.00000             3.864770e+05
75%             702.00000             8.093410e+05
max            2893.00000             3.844585e+06

In [136]:
potential_areas_df['state'].value_counts()

California    149
Oregon        130
Arizona        76
Nevada         48
Washington     38
Name: state, dtype: int64

In [137]:
potential_areas_df['gender'].value_counts()

M    222
F    219
Name: gender, dtype: int64

In [138]:
potential_areas_df['education'].value_counts()

Bachelor                135
College                 133
High School or Below    129
Master                   27
Doctor                   17
Name: education, dtype: int64

In [139]:
potential_areas_df['policy_type'].value_counts()

Personal Auto     311
Corporate Auto    102
Special Auto       28
Name: policy_type, dtype: int64

In [141]:
potential_areas_df['vehicle_class'].value_counts()

Four-Door Car    209
SUV               96
Two-Door Car      78
Luxury            58
Name: vehicle_class, dtype: int64

### Insights:
1. Most of potential customers comes from California (with 149 out of 441 records. ~33.8%), followed by one coming from Oregon (130, ~29.5%), Arizona (76, ~17.2%), Nevada (48, ~10.9%), and the least from Washington(38, ~8.6%)

2. there is no major different between female and male customers in this potential group

3. Majority of cusomters has an education of Bacherlor (135, ~30.6%), College (133, ~30.2%), and High school or below (129, ~29.3%); while very few customers has a higher education master (27, ~6.1%) or doctor (17, ~3.9%).

4. most of customers has a personal auto policy type (311 out 441,~70.5%) followed by corporate auto policy (102,, ~23.1%), and special auto policy (28,~6.3%) 

5. most of these customers has four-Door car (209, ~47.4%), followed by SUV (96, ~21.8%), two-Door car (78, ~ 17.7%), then luxury (58, ~13.2%)